File: python/isomorphic_strings.py
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        """time complexity: O(n), space complexity: O(n)"""
        if len(s) != len(t):
            return False
        s_t_map: dict[str, str] = {}
        t_s_map: dict[str, str] = {}
        for i in range(len(s)):
            if s[i] not in s_t_map:
                s_t_map[s[i]] = t[i]
            elif s_t_map[s[i]] != t[i]:
                return False
            if t[i] not in t_s_map:
                t_s_map[t[i]] = s[i]
            elif t_s_map[t[i]] != s[i]:
                return False
        return True


def test_isIsomorphic_case_1():
    # arrange
    s: str = "egg"
    t: str = "add"
    expected: bool = True

    # act
    solution = Solution()
    actual = solution.isIsomorphic(s, t)

    # assert
    assert expected == actual


def test_isIsomorphic_case_2():
    # arrange
    s: str = "foo"
    t: str = "bar"
    expected: bool = False

    # act
    solution = Solution()
    actual = solution.isIsomorphic(s, t)

    # assert
    assert expected == actual


def test_isIsomorphic_case_3():
    # arrange
    s: str = "paper"
    t: str = "title"
    expected: bool = True

    # act
    solution = Solution()
    actual = solution.isIsomorphic(s, t)

    # assert
    assert expected == actual


def test_isIsomorphic_case_4():
    # arrange
    s: str = "badc"
    t: str = "baba"
    expected: bool = False

    # act
    solution = Solution()
    actual = solution.isIsomorphic(s, t)

    # assert
    assert expected == actual

File: rust/src/code/isomorphic_strings.rs
struct Solution {}

impl Solution {
    // NOTE: time complexity: O(n)
    pub fn is_isomorphic(s: String, t: String) -> bool {
        use std::collections::HashMap;

        let s: Vec<char> = s.chars().collect();
        let t: Vec<char> = t.chars().collect();
        if s.len() != t.len() {
            return false;
        }
        let mut s_t_map: HashMap<char, char> = HashMap::new();
        let mut t_s_map: HashMap<char, char> = HashMap::new();
        for i in 0..s.len() {
            if let Some(c) = s_t_map.get(&s[i]) {
                if *c != t[i] {
                    return false;
                }
            } else {
                s_t_map.insert(s[i], t[i]);
            }
            if let Some(c) = t_s_map.get(&t[i]) {
                if *c != s[i] {
                    return false;
                }
            } else {
                t_s_map.insert(t[i], s[i]);
            }
        }
        true
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_is_isomorphic_case_1() {
        // arrange
        let s: String = "egg".to_string();
        let t: String = "add".to_string();
        let expected: bool = true;

        // act
        let actual = Solution::is_isomorphic(s, t);

        // assert
        assert_eq!(expected, actual);
    }
}
